File: OrangeJuice.Server.Core/Web/EncodedQueryBuilder.cs
using System;
using System.Collections.Generic;

namespace OrangeJuice.Server.Web
{
	public sealed class EncodedQueryBuilder : IQueryBuilder
	{
		#region Fields
		private readonly IUrlEncoder _urlEncoder;
		#endregion

		#region Ctor
		public EncodedQueryBuilder(IUrlEncoder urlEncoder)
		{
			_urlEncoder = urlEncoder;
		}
		#endregion

		#region IQueryBuilder members
		public string BuildQuery(IEnumerable<KeyValuePair<string, string>> args)
		{
			var coll = HttpUtility.ParseQueryString(String.Empty, _urlEncoder);
			foreach (var arg in args)
			{
				coll.Add(arg.Key, arg.Value);
			}
			return coll.ToString();
		}

		public string SignQuery(string query, string signature)
		{
			var coll = HttpUtility.ParseQueryString(query, _urlEncoder);
			coll.Add("Signature", signature);
			return coll.ToString();
		}
		#endregion
	}
}

File: OrangeJuice.Server.Core.FSharp/Web/EncodedQueryBuilder.fs
namespace OrangeJuice.Server.FSharp.Web

open OrangeJuice.Server.Web

open System
open System.Collections.Generic

type EncodedQueryBuilder(urlEncoder : IUrlEncoder) =
    interface IQueryBuilder with
        member this.BuildQuery(args : IEnumerable<KeyValuePair<string, string>>) : string =
            let coll = HttpUtility.ParseQueryString(String.Empty, urlEncoder)
            args |> Seq.iter (fun arg -> coll.Add(arg.Key, arg.Value))
            coll.ToString()

        member this.SignQuery(query : string, signature : string) =
            let coll = HttpUtility.ParseQueryString(query, urlEncoder)
            coll.Add("Signature", signature) |> ignore
            coll.ToString()
